designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+       is deprecated".
+    2. An _action_ the user should take, such as "Use = instead". If stanc3 can
+       automate this for them, we should call that out, leading to a full
+       message like "Use = instead or run stanc3 --print-canonical to update
+       print an updated version of your code."
+    3. A timeline which includes a date or version after which this feature is
+       slated for removal, for example "This deprecation will expire in Stan
+       version 2.31.0 (expected Sept. 2022)".
+2. It has been added to the [documentation on
+   deprecation](https://mc-stan.org/docs/reference-manual/deprecated-features-appendix.html)
+   with the same 3 pieces of information, and additional information as needed
+   (such as any of: reasoning for removal, advantages of new behavior, ongoing
+   related changes, etc).
+3. Examples in the documentation have been updated to remove any examples of the
+   deprecated syntax, except for those referring to its deprecation. This change
+   may lag behind bullets 1 and 2, but must occur before the removal.
+
+Our goal is to only deprecate a feature when doing so provides a tangible
+benefit to users, such as improving error messaging or allowing new features to
+be added to the language while keeping a low maintenance burden. Changes should
+never take users (even those with many or large models) long to adjust to, and
+as few such changes should be made in each version as is reasonable.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Developers
+From a developer standpoint, this policy is rather straightforward. When a
+change is made, it should identify any deprecation candidates: new keywords that
+must be implemented (thus 'deprecating' the use of that word as a identifier),
+new syntax which supersedes prior syntax, (such as `=` or `array[]`), or a
+similar **syntax-level** change. If supporting both the new and old way of doing
+things requires a 'hack' or difficult to maintain code, this is a prime
+candidate for this deprecation-removal process. If a change meaningfully changes
+the semantics of the language, it is not.
+
+>Current examples include the use of 'partial keywords' like offset,
+>which are not reserved words but act like them in some contexts, or multiple
+>syntaxes which lead to parsing difficulties, like the new array syntax. We would
+>like to not only improve the error messaging and implementation of these current
+>features, but also avoid introducing new issues of this variety in the long term.
+
+>Current non-examples include the ODE solver interface changes, which requires
+>changes to the user-defined function defining the ODE.
+
+Once it has been decided to deprecate a feature, the lexer, parser, or
+typechecker (whichever is most apropriate) should be updated to emit a
+compile-time warning notifying the user. This must contain the above 3 criteria:
```

Comment:
I'd strongly recommend using "this" less and repeating nouns more.  So rather than "This" I'd say "The warning" here.  

But the real issue is that these aren't criteria.  Small numbers used as adjectives should be spelled out, so it's "three" not "3" here.  So overall, I'd go with

```
The warning message must specify three pieces of information, (a) what
is deprecated, (b) how to adjust code to replace the deprecated feature, and (c) the
version in which the deprecated feature will be removed.
```